File: kupas/api/schemas/extraction.py
"""
kupas/api/schemas/extraction.py
Pydantic v2 schemas for the PDF Extraction & Verification API.
"""

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class ExtractionChunkCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    content: str = Field(..., min_length=100)
    start_page: int = Field(..., ge=1)
    end_page: int = Field(..., ge=1)
    is_verified: bool = False


class ExtractionChunkOut(BaseModel):
    id: int
    session_id: int
    title: str
    content: str
    start_page: int
    end_page: int
    char_count: int
    word_count: int
    quality_score: float
    is_verified: bool
    order_index: int
    model_config = {"from_attributes": True}


class ExtractionChunkUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    content: Optional[str] = Field(None, min_length=1)
    is_verified: Optional[bool] = None


class ExtractionSessionOut(BaseModel):
    id: int
    book_id: int
    status: str
    total_pages: Optional[int]
    approved_by: Optional[int]
    chunks: List[ExtractionChunkOut] = []
    model_config = {"from_attributes": True}


class ExtractionPreviewResponse(BaseModel):
    chunks: List[ExtractionChunkCreate]
    statistics: Dict[str, Any]
    warnings: List[str]


class SaveExtractionRequest(BaseModel):
    book_id: int = Field(..., ge=1)
    chunks: List[ExtractionChunkCreate]
    total_pages: Optional[int] = None

File: frontend-next/types/extraction.ts
export interface ExtractionChunk {
  id?: number
  title: string
  content: string
  start_page: number
  end_page: number
  char_count?: number
  word_count?: number
  quality_score?: number
  is_verified: boolean
  order_index?: number
}

export interface ExtractionStatistics {
  total_chunks: number
  total_chars: number
  total_words: number
  avg_chunk_size: number
  min_chunk_size: number
  max_chunk_size: number
  chunks_below_minimum: number
  chunks_above_maximum: number
}

export interface ExtractionPreviewData {
  chunks: ExtractionChunk[]
  statistics: ExtractionStatistics
  warnings: string[]
}

export interface ExtractionSession {
  id: number
  book_id: number
  status: 'draft' | 'in_review' | 'approved' | 'rejected'
  total_pages: number | null
  approved_by: number | null
  chunks: ExtractionChunk[]
}
